p2p/src/peer_manager/mod.rs

```diff
@@ -115,50 +129,45 @@ where
         !self.peerdb.is_id_banned(peer_id)
     }
 
-    /// Verify protocol compatibility
+    /// Verifies the protocols compatibility.
     ///
-    /// Make sure that remote peer supports the same versions of the protocols that we do
-    /// and that they support the mandatory protocols which for now are configured to be:
+    /// Checks that the given set of protocols contains the following protocols with the exact
+    /// versions:
     ///
-    /// - `/meshsub/1.1.0`
-    /// - `/meshsub/1.0.0`
-    /// - `/ipfs/ping/1.0.0`
-    /// - `/ipfs/id/1.0.0`
-    /// - `/ipfs/id/push/1.0.0`
-    /// - `/mintlayer/sync/0.1.0`
+    /// - PubSub 1.1.0
+    /// - Ping 1.0.0
+    /// - Sync 0.1.0
     ///
-    /// If any of the procols are missing or if any of them have a different version,
+    /// If any of the protocols are missing or if any of them have a different version,
     /// the validation fails and connection must be closed.
     ///
     /// Either peer may support additional protocols which are not known to the other
     /// peer and that is totally fine. As long as the aforementioned protocols with
     /// matching versions are found, the protocol set has been validated successfully.
-    // TODO: create generic versions of the protocols when mock interface is supported again
-    // TODO: convert `protocols` to a hashset
-    // TODO: define better protocol id type
-    fn validate_supported_protocols(&self, protocols: &[T::ProtocolId]) -> bool {
-        const REQUIRED: &[&str] = &[
-            "/meshsub/1.1.0",
-            "/meshsub/1.0.0",
-            "/ipfs/ping/1.0.0",
-            "/ipfs/id/1.0.0",
-            "/ipfs/id/push/1.0.0",
-            "/mintlayer/sync/0.1.0",
+    fn validate_supported_protocols(&self, protocols: &HashSet<Protocol>) -> bool {
```

Comment:
So. Here's the other problem I wanted to talk about. Usually protocol comparison is upgrade-based. Meaning, the general case is expected compatibility, not incompatibility. Let's say we support 5 protocols. All have 5 names and 5 versions. Now the idea that we have 5 protocol versions is a big issue already. There should be one "Mintlayer protocol version". If we upgrade one of the five protocols, we shouldn't check for the compatibility of each individual protocol, but only for the Mintlayer upgrade. So if we upgrade from version 0.1.0 to 0.2.0, we check that the peer has `0.1.0 <= version <= 0.2.0`. That's much more maintainable that having 5 different versions.

Now I understand that this whole 5 different versions thing is inherent from the Behaviour thing in libp2p... and I don't know how to solve this problem. But maybe we should have a discussion to see how this should work. This whole substream thing is becoming messy very quickly.